File: src/akgentic/agent_card.py
"""Agent profile cards for team capability discovery.

AgentCards describe available agent profiles/roles in a team, enabling
dynamic discovery of capabilities and agent creation patterns.

IMPORTANT: Always use get_config() to obtain config for agent instantiation.
This ensures each agent gets an independent copy, preventing shared mutable state bugs."""

from __future__ import annotations

from typing import Any

from akgentic.agent_config import BaseConfig
from akgentic.utils.serializer import SerializableBaseModel


class AgentCard(SerializableBaseModel):
    """Describes an agent profile/role available in the team.

    AgentCards form a capability catalog, allowing agents to discover
    what profiles exist and how to work with them. This is not instance
    tracking—it's a profile directory.

    Example:
        >>> card = AgentCard(
        ...     role="ResearchAgent",
        ...     description="Performs web research and data gathering",
        ...     skills=["web_search", "pdf_extraction"],
        ...     agent_class="examples.multi_agent.ResearchAgent",
        ...     config=BaseConfig(name="research", role="ResearchAgent"),
        ...     routes_to=["WriterAgent", "AnalystAgent"]
        ... )
        >>> # Now other agents can discover this profile and its config
        >>> print(card.skills)
        ['web_search', 'pdf_extraction']
        >>> card.can_route_to("WriterAgent")
        True

    Attributes:
        role: Agent role/type identifier (e.g., "ResearchAgent")
        description: Human-readable description of what this agent does
        skills: List of capabilities this agent provides
        agent_class: Fully qualified class name (str) or actual class (type) for instantiation
        config: Default BaseConfig (or subclass) for this profile
        routes_to: List of roles this agent can send requests to.
                   Empty list means can route to any role (no restrictions).
                   Agents can always respond to requests regardless of this field.
        metadata: Extensible key-value storage for custom attributes
    """

    role: str
    description: str
    skills: list[str]
    agent_class: str | type
    config: BaseConfig | dict[str, Any] = {}
    routes_to: list[str] = []
    metadata: dict[str, Any] = {}

    def get_config_copy(self) -> BaseConfig:
        """Get a deep copy of the config as BaseConfig instance.

        **ALWAYS use this method when creating agents from an AgentCard.**
        Returns a fresh copy to prevent shared mutable state across multiple
        agent instances created from the same AgentCard.

        Returns:
            Deep copy of BaseConfig instance (converts dict if needed)

        Example:
            >>> card = orchestrator.get_agent_card("Developer")
            >>> config = card.get_config()  # Safe - returns independent copy
            >>> config.name = "alice"  # Won't affect other agents
        """

        if isinstance(self.config, BaseConfig):
            return self.config.model_copy(deep=True)

        return BaseConfig(**self.config)

    def has_skill(self, skill: str) -> bool:
        """Check if this profile has a specific skill.

        Args:
            skill: Skill to check for

        Returns:
            True if skill is in the profile's skill list
        """
        return skill in self.skills

    def can_route_to(self, role: str) -> bool:
        """Check if this profile can send requests to a specific role.

        An empty routes_to list means no restrictions (can route to anyone).
        Otherwise, the target role must be in the routes_to list.

        Note: Agents can always RESPOND to requests from any role.
        This only controls which roles an agent can proactively SEND to.

        Args:
            role: Target role to check routing permission for

        Returns:
            True if this agent profile can send requests to the target role
        """
        if not self.routes_to:
            return True  # No restrictions - can route to any role
        return role in self.routes_to

File: lib/akgentic/agent_card.ex
defmodule Akgentic.AgentCard do
  @moduledoc """
  Agent profile cards for team capability discovery.

  AgentCards describe available agent profiles/roles in a team, enabling
  dynamic discovery of capabilities and agent creation patterns.

  ## Python → Elixir Mapping

  Maps from Python's `AgentCard(SerializableBaseModel)` to an Elixir struct
  with equivalent fields and functions.

  ## Examples

      card = Akgentic.AgentCard.new(
        role: "ResearchAgent",
        description: "Performs web research and data gathering",
        skills: ["web_search", "pdf_extraction"],
        agent_module: MyApp.ResearchAgent,
        config: %{name: "research", role: "ResearchAgent"},
        routes_to: ["WriterAgent", "AnalystAgent"]
      )

      card.skills
      # => ["web_search", "pdf_extraction"]

      Akgentic.AgentCard.has_skill?(card, "web_search")
      # => true

      Akgentic.AgentCard.can_route_to?(card, "WriterAgent")
      # => true
  """

  @type t :: %__MODULE__{
          role: String.t(),
          description: String.t(),
          skills: [String.t()],
          agent_module: module() | String.t(),
          config: map(),
          routes_to: [String.t()],
          metadata: map()
        }

  @enforce_keys [:role, :description, :skills, :agent_module]
  defstruct [
    :role,
    :description,
    :skills,
    :agent_module,
    config: %{},
    routes_to: [],
    metadata: %{}
  ]

  @doc """
  Create a new AgentCard.

  ## Options

    * `:role` - Agent role/type identifier (required)
    * `:description` - Human-readable description (required)
    * `:skills` - List of capabilities (required)
    * `:agent_module` - Module or string for agent class (required)
    * `:config` - Default configuration map (default: %{})
    * `:routes_to` - List of roles this agent can send to (default: [])
    * `:metadata` - Extensible key-value storage (default: %{})

  ## Examples

      card = Akgentic.AgentCard.new(
        role: "Worker",
        description: "Processes tasks",
        skills: ["compute"],
        agent_module: MyApp.WorkerAgent
      )
  """
  @spec new(keyword()) :: t()
  def new(opts) do
    struct!(__MODULE__, opts)
  end

  @doc """
  Get a deep copy of the config map.

  Always use this method when creating agents from an AgentCard
  to prevent shared mutable state.

  ## Examples

      config = Akgentic.AgentCard.get_config_copy(card)
  """
  @spec get_config_copy(t()) :: map()
  def get_config_copy(%__MODULE__{config: config}) do
    # Deep copy via serialization round-trip
    config
    |> :erlang.term_to_binary()
    |> :erlang.binary_to_term()
  end

  @doc """
  Get the agent module as a module atom.

  If `agent_module` is a string, it is converted to a module atom.

  ## Examples

      module = Akgentic.AgentCard.get_agent_module(card)
  """
  @spec get_agent_module(t()) :: module()
  def get_agent_module(%__MODULE__{agent_module: module}) when is_atom(module), do: module

  def get_agent_module(%__MODULE__{agent_module: module_string}) when is_binary(module_string) do
    module_string
    |> String.split(".")
    |> Enum.map(&String.to_existing_atom/1)
    |> Module.concat()
  end

  @doc """
  Check if this profile has a specific skill.

  ## Examples

      Akgentic.AgentCard.has_skill?(card, "web_search")
      # => true
  """
  @spec has_skill?(t(), String.t()) :: boolean()
  def has_skill?(%__MODULE__{skills: skills}, skill) do
    skill in skills
  end

  @doc """
  Check if this profile can send requests to a specific role.

  An empty `routes_to` list means no restrictions (can route to anyone).
  Otherwise, the target role must be in the `routes_to` list.

  Note: Agents can always respond to requests from any role.
  This only controls which roles an agent can proactively send to.

  ## Examples

      Akgentic.AgentCard.can_route_to?(card, "WriterAgent")
      # => true
  """
  @spec can_route_to?(t(), String.t()) :: boolean()
  def can_route_to?(%__MODULE__{routes_to: []}, _role), do: true
  def can_route_to?(%__MODULE__{routes_to: routes_to}, role), do: role in routes_to

  @doc """
  Serialize the agent card to a map.

  ## Examples

      map = Akgentic.AgentCard.to_map(card)
  """
  @spec to_map(t()) :: map()
  def to_map(%__MODULE__{} = card) do
    agent_module =
      case card.agent_module do
        module when is_atom(module) -> inspect(module)
        string when is_binary(string) -> string
      end

    %{
      role: card.role,
      description: card.description,
      skills: card.skills,
      agent_module: agent_module,
      config: card.config,
      routes_to: card.routes_to,
      metadata: card.metadata
    }
  end

  @doc """
  Deserialize an agent card from a map.

  ## Examples

      card = Akgentic.AgentCard.from_map(map)
  """
  @spec from_map(map()) :: t()
  def from_map(map) when is_map(map) do
    %__MODULE__{
      role: Map.fetch!(map, :role),
      description: Map.fetch!(map, :description),
      skills: Map.get(map, :skills, []),
      agent_module: Map.fetch!(map, :agent_module),
      config: Map.get(map, :config, %{}),
      routes_to: Map.get(map, :routes_to, []),
      metadata: Map.get(map, :metadata, %{})
    }
  end
end
